compiler/compiler-core.cpp

```diff
@@ -494,6 +494,14 @@ const Index &CompilerCore::get_index() {
 }
 
 File *CompilerCore::get_file_info(std::string &&file_name) {
+  File *f = cpp_index.get_file(file_name);
```

Comment:
Here a race condition with `insert_file` call is possible.

Reply:
Well... It's funny, that this race condition existed earlier (see WriteFilesF in the current master). But somehow, we didn't encounter it in a real life. Why? I don't know. Maybe, its probability is too small (e.g. the number of newly-appearing files is negligibly small compared to the number of files from a previous launch). Maybe, it was just a luck.

I couldn't find the way to fix this simply. That's why I have written a complicated fix, see the last commit. 
Its main idea is the following:
* don't keep a single map<string, File*> files:
* instead, have two maps: files_prev_launch and files_only_cur_launch
* the first map is accessed in a single thread (when reading from dir / indexfile)
* the second is accessed from multiple threads, that's why it's guarded with a mutex
Please, consider the last commit and comments within it.